staging/src/k8s.io/apiserver/pkg/storage/cacher/store_btree.go

```diff
@@ -0,0 +1,404 @@
+/*
+Copyright 2022 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cacher
+
+import (
+	"fmt"
+	"strings"
+	"sync"
+
+	"github.com/google/btree"
+	"k8s.io/client-go/tools/cache"
+)
+
+func newThreadedBtreeStoreIndexer(indexers cache.Indexers, degree int) *threadedStoreIndexer {
+	return &threadedStoreIndexer{
+		store:   newBtreeStore(degree),
+		indexer: newIndexer(indexers),
+	}
+}
+
+type threadedStoreIndexer struct {
```

Comment:
What's the point of having a separate implementation that only differes by adding locking?

Reply:
For snapshotting needed to serve continuations and exact match in https://github.com/kubernetes/kubernetes/pull/126855. Snapshotting btree is cheap, while snapshotting golang map requires copying out all the contents. So I decided on a following structure:

`threadedStoreIndexer` as a thread safe combination of store and index to keep latest state and serve latest revision reads.
When needed we snapshot the store (cheap) and serve reads for older revisions. No locking needed as we can make the data immutable.

Reply:
Please add a comment explaining that here.